Branch-SDK/src/main/java/io/branch/referral/BranchPostTask.java

```diff
@@ -0,0 +1,181 @@
+package io.branch.referral;
+
+import androidx.annotation.NonNull;
+
+import org.json.JSONException;
+import org.json.JSONObject;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.concurrent.CountDownLatch;
+
+import static io.branch.referral.BranchError.ERR_BRANCH_REQ_TIMED_OUT;
+
+/**
+ * Asynchronous task handling execution of server requests. Execute the network task on background
+ * thread and request are  executed in sequential manner. Handles the request execution in
+ * Synchronous-Asynchronous pattern. Should be invoked only form main thread and  the results are
+ * published in the main thread.
+ */
+public class BranchPostTask extends BranchAsyncTask<Void, Void, ServerResponse> {
```

Comment:
this is a massive class that used to live inside the Branch.java singleton. I was able to pass the SDK singleton as constructor argument and make the relevant methods & parameters package private, so moving this class out to its own file should not have an effect.

However, the `onPostExecute` method changed a little bit, it would be best to look at the original and compare the two. In a nutshell, I could not understand why the requests (once completed) were being removed from the request queue **by index** and not **by reference** to the completed request, which is just so much safer. In addition, we used to remove ALL requests from the queue once we encounter the first failed request whose status code is not 400/409. I disagree with the assumption that all subsequent requests will fail. With the flaky connections on mobile phones being so common, the next request could succeed just fine... To that end, I now only remove the recently completed request from the queue and leave all other requests on the queue.